### PR TITLE
Fix Psalm-issues batch #3

### PR DESCRIPTION
**Pull request by tvdijen**
Fixes all Psalm-issues in all if `SimpleSAML/lib/`.

### CODE REVIEW

File: lib/SimpleSAML/IdP.php [hunk 1]
```diff
@@ -505,8 +505,10 @@ public function handleLogoutRequest(array &$state, $assocId)
 
         $this->authSource->logout($returnTo);
 
-        $handler = $this->getLogoutHandler();
-        $handler->startLogout($state, $assocId);
+        if ($assocId !== null) {
```
Comment (tvdijen) [inline lib/SimpleSAML/IdP.php]: Twelve lines above we also test for `$assocId` being `null`, so I'm guessing this could be a valid situation

File: lib/SimpleSAML/IdP.php [hunk 2]
```diff
@@ -526,8 +528,11 @@ public function handleLogoutResponse($assocId, $relayState, Error\Exception $err
         assert(is_string($assocId));
         assert(is_string($relayState) || $relayState === null);
 
+        $index = strpos($assocId, ':');
```
Comment (tvdijen) [inline lib/SimpleSAML/IdP.php]: Perhaps raising an exception if `$assocId` does not meet a certain format would be better.. But what would that format be?

File: lib/SimpleSAML/Logger.php [hunk 3]
```diff
@@ -13,9 +13,14 @@
...
-     * @var \SimpleSAML\Logger\LoggingHandlerInterface|false|null
+     * @var \SimpleSAML\Logger\LoggingHandlerInterface
      */
-    private static $loggingHandler = null;
+    private static $loggingHandler;
+
+    /**
+     * @var bool
+     */
+    private static $initializing = false;
```
Comment (tvdijen) [inline lib/SimpleSAML/Logger.php]: By introducing a new property, we can keep the typedecl of `$loggingHandler` as simple as possible..

File: lib/SimpleSAML/Store.php [hunk 4]
```diff
@@ -64,9 +64,11 @@ public static function getInstance()
                         $c
                     );
                 }
+                /** @var \SimpleSAML\Store|false */
                 self::$instance = new $className();
         }
 
+        /** @var \SimpleSAML\Store|false */
```
Comment (jaimeperez) [inline lib/SimpleSAML/Store.php]: Apparently this is not necessary.

File: lib/SimpleSAML/Logger.php [hunk 5]
```diff
@@ -393,6 +398,9 @@ private static function defer($level, $message, $stats)
             register_shutdown_function([self::class, 'shutdown']);
             self::$shutdownRegistered = true;
         }
+
+        // Set initializing to true > all went well!
+        self::$initializing = true;
```
Comment (jaimeperez) [inline lib/SimpleSAML/Logger.php]: There is something weird here, and it's difficult to follow the code with the current name of the property and the way it's assigned values. I think `$initializing` should default to `false`, and set to `true` when we call `createLoggingHandler()`, and back to `false` after we're done there.

File: lib/SimpleSAML/Logger.php [hunk 6]
```diff
@@ -453,20 +461,20 @@ private static function createLoggingHandler($handler = null)
      */
     private static function log($level, $string, $statsLog = false)
     {
-        if (self::$loggingHandler === false) {
+        if (self::$initializing === false) {
```
Comment (jaimeperez) [inline lib/SimpleSAML/Logger.php]: And then this should just be `if (self::$initializing)`, as that's what this code is aiming to capture, the case where we're trying to log something while initializing the logging handler.

File: lib/SimpleSAML/Memcache.php [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 namespace SimpleSAML;
 
+use Memcache as _Memcache;
+use Memcached as _Memcached;
```
Comment (jaimeperez) [inline lib/SimpleSAML/Memcache.php]: These aliases are a bit confusing, and I think we can live without them.

File: lib/SimpleSAML/Memcache.php [hunk 8]
```diff
@@ -304,24 +306,27 @@ private static function addMemcacheServer($memcache, $server)
...
                 'Missing Memcached implementation. You must install either the Memcache or Memcached extension.'
             );
-        } elseif (strtolower($class) === '\memcache') {
-            Logger::warning("The use of PHP-extension memcache is deprecated. Please migrate to the memcached extension.");
         }
-        self::$extension = strtolower($class);
 
-        // create the \Memcache object
-        $memcache = new $class();
+        if (self::$extension === '\memcache') {
```
Comment (jaimeperez) [inline lib/SimpleSAML/Memcache.php]: Is there a reason for using lowercase here? Can't we just compare with `\Memcache`?

File: lib/SimpleSAML/SessionHandlerPHP.php [hunk 9]
```diff
@@ -364,7 +370,7 @@ public function setCookie($sessionName, $sessionID, array $cookieParams = null)
             );
         }
 
-        session_id($sessionID);
+        session_id(strval($sessionID));
```
Comment (jaimeperez) [inline lib/SimpleSAML/SessionHandlerPHP.php]: This sounds unnecessary?
